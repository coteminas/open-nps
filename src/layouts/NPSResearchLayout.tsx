import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import get from 'lodash.get';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import {
  ThemeProvider,
  createMuiTheme,
  ThemeOptions,
  makeStyles,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';

import { ITarget } from '~/model/Target';
import { IResearch } from '~/model/Research';
import { IConfig } from '~/model/Config';
import { IReviewer } from '~/model/Reviewer';
import { AddThemeOptsDefaults } from '~/util/themeOpts';

const styles = (theme) =>
  createStyles({
    root: {
      height: '100vh',
      maxWidth: '750px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      padding: theme.spacing(2),

      '& > h4': {
        marginBottom: theme.spacing(3),
        textAlign: 'justify',
      },
    },
  });

const useStyles = makeStyles(styles);

export interface LayoutProps extends LayoutSharedProps {
  layoutClasses: WithStyles<ReturnType<typeof styles>>['classes'];
}

interface LayoutSharedProps {
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  researchId: string;
  data: {
    reviewer: IReviewer['meta'];
    target: IReviewer['meta'];
  };
}

interface Props extends LayoutSharedProps {
  mui: ThemeOptions;
}

interface GetServerSideProps {
  ctxResearchIdGetter(ctx: GetServerSidePropsContext): string;
  researchExtraData: { concluded: boolean };
}
export const getServerSidePropsFn = ({
  ctxResearchIdGetter,
  researchExtraData,
}: GetServerSideProps) => async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const researchId = ctxResearchIdGetter(ctx);
  const { connectMongo } = await import('../util/mongo');
  const { Target, Research } = await import('../model');

  await connectMongo();

  const research: IResearch = await Research.findOne({
    _id: researchId,
    ...researchExtraData,
  }).populate('reviewer', { _id: 0 });

  const target: ITarget = research
    ? await Target.findById(research.target).populate('configs', { _id: 0 })
    : null;

  if (!research || !target) {
    return { notFound: true };
  }

  const configs: AnyObject = (target.configs as IConfig[]).reduce(
    (a, c) => ({ ...a, [c.key]: c.values }),
    {}
  );

  return {
    props: {
      mui: configs.mui,
      templates: configs.templates,
      themeOpts: AddThemeOptsDefaults(configs.themeOpts),
      data: {
        reviewer: get(research, 'reviewer.meta', {}),
        target: get(target, 'meta', {}),
      },
      researchId,
    },
  };
};

/* istanbul ignore next */
export const removeLocalJSS = (): void => {
  const jssStyles = document.querySelector('#jss-server-side');
  if (jssStyles) {
    jssStyles.parentElement.removeChild(jssStyles);
  }
};

export const withLayout = (
  Component: React.FC<LayoutProps>
): React.FC<Props> => ({ mui, ...props }): React.ReactElement => {
  const classes: LayoutProps['layoutClasses'] = useStyles();

  React.useEffect(removeLocalJSS, []);

  return (
    <ThemeProvider theme={createMuiTheme(mui)}>
      <CssBaseline />
      <Component {...props} layoutClasses={classes} />
    </ThemeProvider>
  );
};

export default withLayout;
