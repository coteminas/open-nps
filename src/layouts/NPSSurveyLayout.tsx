import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import get from 'lodash.get';
import merge from 'lodash.merge';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import {
  ThemeProvider,
  createMuiTheme,
  ThemeOptions,
  makeStyles,
  WithStyles,
  createStyles,
  Theme,
} from '@material-ui/core/styles';

import { ITarget } from '~/model/Target';
import { ISurvey } from '~/model/Survey';
import { IConfig } from '~/model/Config';
import { AddThemeOptsDefaults } from '~/util/themeOpts';
import { AddDefaultTemplates } from '~/util/templates';

const childrenStyles = ({
  SurveyBoxBorderSize,
  SurveyBoxBorderRadius,
}: ThemeOptionsConfigValues) => (theme: Theme) =>
  createStyles({
    root: {
      maxWidth: '750px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      padding: theme.spacing(4),
      border: `${SurveyBoxBorderSize} solid ${theme.palette.primary.main}`,
      borderRadius: SurveyBoxBorderRadius,
    },
    corePhrase: {
      marginBottom: theme.spacing(3),
      textAlign: 'justify',
    },
    commentPhrase: {
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(1),
      textAlign: 'justify',
    },
    brand: {
      width: '100%',
      marginBottom: theme.spacing(3),

      '& > img': {
        width: '100%',
      },
    },
  });

const styles = (theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      height: '100vh',
      padding: theme.spacing(2),
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    },
  });

const useStyles = makeStyles(styles);

export interface LayoutProps extends LayoutSharedProps {
  layoutClasses: WithStyles<ReturnType<typeof childrenStyles>>['classes'];
}

interface LayoutSharedProps {
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  surveyId: string;
  data: {
    reviewer: AnyObject;
    target: ITarget['meta'];
  };
}

interface Props extends LayoutSharedProps {
  mui: ThemeOptions;
}

interface GetServerSideProps {
  ctxSurveyIdGetter(ctx: GetServerSidePropsContext): string;
  surveyExtraData: { concluded: boolean };
}

const mapConfigsFromList = (configs: IConfig[]) =>
  configs.reduce((a, c) => ({ ...a, [c.key]: c.values }), {});

export const getServerSidePropsFn = ({
  ctxSurveyIdGetter,
  surveyExtraData,
}: GetServerSideProps) => async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const surveyId = ctxSurveyIdGetter(ctx);
  const { connectMongo } = await import('../util/mongo');
  const { Target, Survey } = await import('../model');

  await connectMongo();

  const survey: ISurvey = await Survey.findOne({
    _id: surveyId,
    ...surveyExtraData,
  });

  const target: ITarget = survey
    ? await Target.findById(survey.target).populate('configs', { _id: 0 })
    : null;

  if (!survey || !target) {
    return { notFound: true };
  }

  const overrideConfigsList = await survey.getOverrideConfigs();
  const overrideConfigs: AnyObject[] = overrideConfigsList.map(
    mapConfigsFromList
  );

  const configs: AnyObject = [
    mapConfigsFromList(target.configs as IConfig[]),
    ...overrideConfigs,
  ].reduce((a, b) => merge(a, b), {});

  return {
    props: {
      mui: get(configs, 'mui', {}),
      templates: AddDefaultTemplates(configs.templates),
      themeOpts: AddThemeOptsDefaults(configs.themeOpts),
      data: {
        reviewer: JSON.parse(JSON.stringify(survey.reviewer)),
        target: get(target, 'meta', {}),
      },
      surveyId,
    },
  };
};

/* istanbul ignore next */
export const removeLocalJSS = (): void => {
  const jssStyles = document.querySelector('#jss-server-side');
  if (jssStyles) {
    jssStyles.parentElement.removeChild(jssStyles);
  }
};

export const withLayout = (
  Component: React.FC<LayoutProps>
): React.FC<Props> => ({ mui, ...props }): React.ReactElement => {
  const layoutClasses: LayoutProps['layoutClasses'] = makeStyles(
    childrenStyles(props.themeOpts)
  )();
  const classes = useStyles();

  React.useEffect(removeLocalJSS, []);

  return (
    <ThemeProvider theme={createMuiTheme(mui)}>
      <CssBaseline />
      <div className={classes.root}>
        <Component {...props} layoutClasses={layoutClasses} />
      </div>
    </ThemeProvider>
  );
};

export default withLayout;
