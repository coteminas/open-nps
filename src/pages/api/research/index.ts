import { NextApiRequest, NextApiResponse } from 'next';

import { createApiHandler } from '~/util/api';
import Research, { IResearch } from '~/model/Research';
import Target, { ITarget } from '~/model/Target';
import Reviewer, { IReviewer } from '~/model/Reviewer';

export const createNewResearch = async (
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> => {
  const { reviewerId, reviewerMeta, targetName } = req.body;
  const target: ITarget = await Target.findOne({ name: targetName });
  const savedReviewer: IReviewer = await Reviewer.findOne({
    uniqueIdentifier: reviewerId,
  });
  const reviewer = !savedReviewer
    ? await new Reviewer({
        uniqueIdentifier: reviewerId,
        meta: reviewerMeta,
      }).save()
    : await savedReviewer.updateOne({
        meta: { ...savedReviewer.meta, ...reviewerMeta },
      });

  const research: IResearch = new Research({
    target: target._id,
    reviewer: reviewer._id,
  });

  return res.json(await research.save());
};

export default createApiHandler({
  POST: createNewResearch,
});
