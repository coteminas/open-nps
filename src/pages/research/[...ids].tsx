import React, { useState } from 'react';
import { useRouter } from 'next/router';
import styled, { ThemeProvider } from 'styled-components';

import { ResearchNotes } from '../../components/ResearchNotes';
import { ResearchComment } from '../../components/ResearchComment';

import { Target, Research } from '../../model';

const Test = styled.h2`
  color: ${props => props.theme.test || '#000'}
`

export const getServerSideProps = async (ctx) => {
  const { connectMongo } = await import('../../util/mongo');
  const [ targetName, researchId ] = ctx.params.ids;

  await connectMongo();

  const target = await Target.findOne({ name: targetName }).lean().populate('configs', { _id: 0 }).lean();
  const search = await Research.findOne({ _id: researchId, target: target._id, concluded: false });

  if (!target || !search) {
    return { notFound: true }
  }

  const configs = target.configs.reduce((a, c) => ({ ...a, [c.key]: c.values }), {});
  return { props: { ...configs, targetName, researchId } };
}

interface Props {
  theme: any, texts: any, targetName: string, researchId: string
}
export default ({ theme, texts, targetName, researchId }: Props) => {
  const [ state, setState ] = useState({ note: null, comment: '' });
  const router = useRouter();
  const setValueForField = (field) => (value) => setState({ ...state, [field]: value });

  const onSubmit = async (e) => {
    e.preventDefault();
    const response = await fetch(`${window.location.origin}/api/research/conclude`, {
      method: 'PUT',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ targetName, researchId, ...state })
    });

    const { ok } = await response.json()

    if (ok) {
      router.push('/research/thanks')
    }
  }

  return (
    <ThemeProvider theme={theme}>
      <form onSubmit={onSubmit}>
        <Test>
          { texts.coreQuestion } {state.note} {state.comment}
        </Test>
        <ResearchNotes setValue={setValueForField('note')}/>
        <ResearchComment value={state.comment} setValue={setValueForField('comment')}/>
        <button>
          Enviar
        </button>
      </form>
    </ThemeProvider>
  )
};;
