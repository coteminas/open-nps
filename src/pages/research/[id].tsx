import React, { useState } from 'react';
import Typography from '@material-ui/core/Typography';

import { useRouter, NextRouter } from 'next/router';
import { GetServerSidePropsContext } from 'next';

import ResearchNotes from '~/components/ResearchNotes';
import ResearchComment from '~/components/ResearchComment';
import ResearchSubmit from '~/components/ResearchSubmit';

import { template } from '~/util/template';
import {
  withLayout,
  LayoutProps,
  getServerSidePropsFn,
} from '~/layouts/NPSResearchLayout';

export const ctxResearchIdGetter = (ctx: GetServerSidePropsContext): string =>
  ctx.params.id as string;

export const getServerSideProps = getServerSidePropsFn({
  ctxResearchIdGetter,
  researchExtraData: { concluded: false },
});

interface SubmitData {
  researchId: string;
  note: string;
  comment: string;
}

export const createSubmit = (data: SubmitData, router: NextRouter) => async (
  e: React.FormEvent<HTMLFormElement>
): Promise<void> => {
  e.preventDefault();
  const response = await fetch(
    `${window.location.origin}/api/research/conclude`,
    {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    }
  );

  const { ok } = await response.json();

  if (ok) {
    router.push(`/research/thanks?researchId=${data.researchId}`);
  }
};

export const setValueForFieldInState = (
  state: AnyObject,
  setState: SimpleFn<AnyObject, void>
) => (field: string) => (value: string): void =>
  setState({ ...state, [field]: value });

export const ResearchPage: React.FC<LayoutProps> = ({
  themeOpts,
  templates,
  data,
  researchId,
  layoutClasses,
}): React.ReactElement => {
  const [state, setState] = useState({ note: null, comment: '' });
  const router = useRouter();
  const onSubmit = createSubmit({ researchId, ...state }, router);
  const setValueForField = setValueForFieldInState(state, setState);

  return (
    <form className={layoutClasses.root} onSubmit={onSubmit}>
      <Typography variant="h2" component="h2">
        {template(templates.CoreQuestionPhrase, data)}
      </Typography>
      <ResearchNotes
        themeOpts={themeOpts}
        setValue={setValueForField('note')}
        selected={state.note}
      />
      <ResearchComment
        value={state.comment}
        setValue={setValueForField('comment')}
        label={templates.ResearchCommentLabel}
        placeholder={templates.ResearchCommentPlaceholder}
      />
      <ResearchSubmit themeOpts={themeOpts}>Enviar</ResearchSubmit>
    </form>
  );
};

export default withLayout(ResearchPage);
