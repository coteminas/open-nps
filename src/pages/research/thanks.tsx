import React from 'react';
import Typography from '@material-ui/core/Typography';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import { ThemeOptions } from '@material-ui/core/styles';
import { withStyles, WithStyles, createStyles } from '@material-ui/styles';

import Research from '~/model/Research';
import Target, { ITarget } from '~/model/Target';

import { AddThemeOptsDefaults } from '~/util/themeOpts';
import { template } from '~/util/template';
import { IReviewer } from '~/model/Reviewer';

const styles = () =>
  createStyles({
    root: {
      maxWidth: '800px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    },
  });

interface Props extends WithStyles<typeof styles> {
  mui: ThemeOptions;
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  researchId: string;
  data: {
    reviewer: IReviewer;
    target: ITarget;
  };
}

export const getServerSideProps = async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const { connectMongo } = await import('../../util/mongo');
  const { researchId } = ctx.query;

  await connectMongo();

  const research = await Research.findOne({ _id: researchId })
    .populate('reviewer', { _id: 0 })
    .lean();

  const target = await Target.findOne({ _id: research.target })
    .populate('configs', { _id: 0 })
    .lean();

  const { themeOpts, ...configs } = target.configs.reduce(
    (a, c) => ({ ...a, [c.key]: c.values }),
    {}
  );
  return {
    props: {
      ...configs,
      themeOpts: AddThemeOptsDefaults(themeOpts),
      researchId,
      data: {
        reviewer: (research.reviewer as IReviewer).meta || {},
        target: target.meta,
      },
    },
  };
};

export const ThanksPage: React.FC<Props> = ({ classes, templates, data }) => (
  <div className={classes.root}>
    <Typography variant="h2" component="h2">
      {template(templates.ThanksPhrase, data)}
    </Typography>
  </div>
);

export default withStyles(styles)(ThanksPage);
