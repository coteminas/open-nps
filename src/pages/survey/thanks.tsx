import React from 'react';
import get from 'lodash.get';
import Typography from '@material-ui/core/Typography';

import { GetServerSidePropsContext } from 'next';

import {
  withLayout,
  getServerSidePropsFn,
  LayoutProps,
} from '~/layouts/NPSSurveyLayout';
import { renderTemplate } from '~/util/renderTemplate';

export const ctxSurveyIdGetter = (ctx: GetServerSidePropsContext): string =>
  ctx.query.surveyId as string;

export const getServerSideProps = getServerSidePropsFn({
  ctxSurveyIdGetter,
  surveyExtraData: { concluded: true },
});

export const ThanksPage: React.FC<LayoutProps> = ({
  layoutClasses,
  templates,
  data,
  themeOpts,
}) => (
  <div className={layoutClasses.root}>
    {themeOpts.ThanksTopImage && (
      <div
        style={{
          maxWidth: get(themeOpts, 'ThanksTopImage.width', 'auto'),
        }}
        className={layoutClasses.brand}
      >
        <img
          alt={themeOpts.ThanksTopImage.alt}
          src={themeOpts.ThanksTopImage.url}
        />
      </div>
    )}
    <Typography data-cy="ThanksPageTypography" variant="h3" component="h3">
      {renderTemplate(templates.ThanksPhrase, data)}
    </Typography>
    {templates.ThanksSubPhrase && (
      <Typography variant="h4" component="h4">
        {renderTemplate(templates.ThanksSubPhrase, data)}
      </Typography>
    )}
  </div>
);

export default withLayout(ThanksPage);
