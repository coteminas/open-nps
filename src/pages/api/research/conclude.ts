import { createApiHandler } from "../../../util/api"
import { NextApiRequest, NextApiResponse } from "next"
import { Target, Research } from "../../../model";
import { IResearch } from "../../../model/Research";

const concludeResearch = async (req: NextApiRequest, res: NextApiResponse) => {
  const { note, comment, researchId } = req.body;

  const research: IResearch = await Research.findOne({ _id: researchId, concluded: false });
  const response = await research.updateOne({ concluded: true, note, comment });

  return res.json(response)
}

export default createApiHandler({
  PUT: concludeResearch
})
