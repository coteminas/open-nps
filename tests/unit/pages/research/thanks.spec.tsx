import React from 'react';

import { ThanksPage, ctxResearchIdGetter } from '~/pages/research/thanks';
import { LayoutProps } from '~/layouts/NPSResearchLayout';
import { shallow } from 'enzyme';
import { GetServerSidePropsContext } from 'next';

describe('/src/pages/research/thanks', () => {
  it('should render properly', () => {
    const props = ({
      templates: {
        ThanksPhrase: 'Foo Bar',
      },
      researchId: 'fake',
      themeOpts: {
        a: 1,
      },
      data: {
        b: 1,
      },
      layoutClasses: {
        root: 'root',
      },
    } as unknown) as LayoutProps;

    const wrap = shallow(<ThanksPage {...props} />);
    expect(wrap).toMatchSnapshot();
  });

  it('ctxResearchIdGetter', () => {
    const ctx = ({
      query: { researchId: 'foo' },
    } as unknown) as GetServerSidePropsContext;
    expect(ctxResearchIdGetter(ctx)).toBe(ctx.query.researchId);
  });
});
