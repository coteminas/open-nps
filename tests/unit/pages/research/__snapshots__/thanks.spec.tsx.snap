// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/src/pages/research/thanks should render properly 1`] = `
<div
  className="root"
>
  <WithStyles(ForwardRef(Typography))
    component="h2"
    variant="h2"
  >
    Foo Bar
  </WithStyles(ForwardRef(Typography))>
</div>
`;
